airflow/executors/knative_executor.py

```diff
@@ -0,0 +1,197 @@
+        """
+        Executes command received and stores result state in queue.
+        :param task_instance:
+        :param key: the key to identify the TI
+        :type key: tuple(dag_id, task_id, execution_date)
+        :param command: the command to execute
+        :type command: str
+        """
+        if key is None:
+            return
+        self.log.info("%s running %s", self.__class__.__name__, task_instance)
+        try:
+            future = make_request_async(
+                task_instance.task_id,
+                task_instance.dag_id,
+                task_instance.execution_date,
+                host=self.host, host_header=self.host_header)
+            resp = await future
+            if resp.status != 200:
+                state = State.FAILED
+                self.log.error("Failed to execute task %s.", str(resp.text))
+                self.result_pipe.send((key, state))
+            else:
+                self.log.info("assuming task success")
+                # self.result_queue.put((key, None))
+                self.pop_running_pipe.send(key)
+
+        except asyncio.InvalidStateError as e:
+            state = State.FAILED
+            self.log.error("Failed to execute task %s.", str(e))
+            self.result_pipe.send((key, state))
+
+
+class TaskMessage:
+    def __init__(self, key, ti):
+        self.key = key
+        self.ti = ti
+
+
+class KnativeExecutor(BaseExecutor):
+    """
+    KnativeExecutor executes tasks locally in parallel. It uses the
+    multiprocessing Python library and queues to parallelize the execution
+    of tasks.
+    """
+    task_queue = None
+    result_queue = None
+    result_pipe = None
+    pop_running_queue = None
+
+    def terminate(self):
+        pass
+
+    @cached_property
+    def kube_executor(self):
+        self.kube_executor_initialized = True
+        return KubernetesExecutor()
+
+    def __init__(self):
+        self.kube_executor_initialized = False
+        super().__init__()
+
+    def start(self):
+        req = conf.get("knative", "knative_host")
+        host_header = None
+        if req is None:
+            raise AirflowException("you must set a knative host")
+
+        self.result_pipe, child_result_pipe = multiprocessing.Pipe()
+        self.task_pipe, child_task_pipe = multiprocessing.Pipe()
+        self.pop_running_pipe, child_pop_running_pipe = multiprocessing.Pipe()
+
+        self.local_loop = KnativeRequestLoop(
+            task_pipe=child_task_pipe,
+            result_pipe=child_result_pipe,
+            host=req,
+            host_header=host_header,
+            pop_running_pipe=child_pop_running_pipe
+        )
+        self.local_loop.start()
+
+    def execute_async(self, key, command, queue=None, executor_config=None, task_instance=None):
+        if queue == "kubernetes":
+            self.kube_executor.execute_async(key, command, queue, executor_config)
+        else:
+            self.task_pipe.send((key, task_instance))
```

Comment:
I think this is going to arbitrarily fail with "object not pickleable", which is why the SimpleTaskInstance object was created in the first place. Is there anything in particular you need from a TI that isn't on a SimpleTaskInstance object?

Reply:
If I'm reading the code right this is being used to pull out `(task_id, dag_id, execution_date)`? If so we already have that. `key` is ` tuple of `dag_id, task_id, execution_date, try_number`